hermes/probe/monitor.go

```diff
@@ -12,9 +12,164 @@
+//	- metricChan: unidirectional channel used for sending metrics to be surfaced.
+func (mp *MonitorProbe) Start(ctx context.Context, metricChan chan<- *metrics.EventMetrics) {
+	probeTicker := time.NewTicker(mp.opts.Interval)
+
+	for {
+		select {
+		case <-ctx.Done():
+			probeTicker.Stop()
+			return
+		case <-probeTicker.C:
+			// init during probe run as metrics are timestamped.
+			if err := mp.initProbeMetrics(); err != nil {
+				mp.Logger.Errorf("error initialising probe metric: %v", err)
+			}
+			mp.runProbe(ctx, metricChan)
+		}
+	}
+}
+
+// initProbeMetrics initializes missing latency metrics for each target.
+// Returns:
+//	- err: returns an error if there is an issue creating a distribution metric from the proto config passed.
+func (mp *MonitorProbe) initProbeMetrics() error {
+	mp.MetricMu.Lock()
+	defer mp.MetricMu.Unlock()
+	for _, t := range mp.targets {
+		if mp.LatencyMetrics[t.GetName()] != nil {
+			continue
+		}
+		var err error
+		if mp.LatencyMetrics[t.GetName()], err = NewProbeMetrics(*mp.Config, *t); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// reportMetrics sends the metrics recorded in the last probe run to Cloudprober.
+// Arguments:
+//	- probeRun: metrics from a probe run on a target.
+//	- metricChan: metric channel passed from Cloudprober.
+func (mp *MonitorProbe) reportMetrics(prm probeRunMetrics, metricChan chan<- *metrics.EventMetrics) {
+	for _, op := range prm.ProbeOpLatency {
+		for _, m := range op {
+			metricChan <- m
+		}
+	}
+
+	for _, call := range prm.ApiCallLatency {
+		for _, m := range call {
+			metricChan <- m
+		}
+	}
+}
+
+// runProbe runs probe for all targets and update EventMetrics.
+// Arguments:
+//	- ctx: pass context to allow for complete cancellation of the probe.
+func (mp *MonitorProbe) runProbe(ctx context.Context, metricChan chan<- *metrics.EventMetrics) {
+	var wg sync.WaitGroup
+	for _, target := range mp.targets {
+		wg.Add(1)
+		go func(target probepb.Target) {
+			defer wg.Done()
+			probeCtx, _ := context.WithDeadline(ctx, time.Now().Add(mp.opts.Timeout))
+			start := time.Now()
+			exitStatus, err := mp.runProbeForTarget(probeCtx, target)
+			if err != nil {
+				mp.Logger.Errorf(err.Error())
+				mp.LatencyMetrics[target.GetName()].ProbeOpLatency["total_probe_run"][exitStatus].Metric("latency").AddFloat64(time.Now().Sub(start).Seconds())
+				return
+			}
+			mp.LatencyMetrics[target.GetName()].ProbeOpLatency["total_probe_run"]["success"].Metric("latency").AddFloat64(time.Now().Sub(start).Seconds())
+			mp.reportMetrics(*mp.LatencyMetrics[target.GetName()], metricChan)
+		}(*target)
+	}
+	wg.Wait()
+}
+
+// runProbeForTarget runs the Hermes probing algorithm on a single target.
+// Arguments:
+//	- ctx: pass context to allow for cancellation of the probe.
+//	- target: the target to be probed
+func (mp *MonitorProbe) runProbeForTarget(ctx context.Context, target probepb.Target) (string, error) {
+	return "", nil
```

Comment:
I would add a TODO or return a "unimplemented" error here.

Reply:
I've added a TODO for now.